### ISSUE TITLE
Feature proposal: customizable text filters for errors

### ISSUE THREAD

**Issue by kwf** [flycheck/flycheck#1479]
Some programming languages, e.g. Rust, usually produce succinct error messages. Others, e.g. Haskell, produce rather verbose errors. When interactively debugging Flycheck errors, I would like to see more succinct messages, even for languages whose compilers are themselves more verbose.

### Proposal

Allow the user to specify a filter which selectively removes unimportant information from the error text reported by a particular specified checker (e.g. `rust-mode` or `haskell-mode`).

### Example

Suppose I write this simple and broken Haskell program, which has a type error:

```haskell
main = putStrLn False
```

The error reported by Flycheck is:

```
[-Wdeferred-type-errors]
     • Couldn't match type ‘Bool’ with ‘[Char]’
       Expected type: String
         Actual type: Bool
     • In the first argument of ‘putStrLn’, namely ‘False’
       In the expression: putStrLn False
       In an equation for ‘main’: main = putStrLn False

 (Hit `C-c C-r' in the Haskell buffer to apply suggestions)
```

To me, the important information in this error is:

```
Couldn't match type ‘Bool’ with ‘[Char]’
Expected type: String
Actual type: Bool
```

In this example, I would want to write a filter which strips all text out of the reported message except for that contained in the first bullet point.

### Additional design notes

Different users might have different preferences for the information they would like to be included by default. For this reason, it makes sense to allow individual users to define filters, rather than leaving that responsibility solely with the author of the checker.

Indeed, all users may wish to see the full error text at some point, so a toggle with a keybinding to temporarily disable these filters would also be desirable.

### Specific benefits

This feature would be useful for many modes of interaction with Flycheck, but it especially would shine in the context of `flycheck-inline-mode` (https://github.com/flycheck/flycheck-inline) or `flycheck-pos-tip-mode` (https://github.com/flycheck/flycheck-pos-tip), as these modes are most useful when the text produced is minimal and to-the-point.

### Implementation thoughts

It may be possible to implement this with existing Flycheck functionality, without modifying Flycheck itself. If so, I'd greatly appreciate a pointer in the right direction. I am a novice elisp programmer, but I am interested in helping to implement this feature, if there is consensus that it would be a good idea.

**Comment by fmdkdd** [comment]
Thank you for the thoughful proposal :+1: 

I agree that Haskell type errors are often long, especially when using `flycheck-inline` or `flycheck-pos-tip`, as you noted.  I would be tempted to argue that GHC needs a `less verbose` flag, but in some cases you do want to see the extra context, so it may be better if the verbosity is actually controlled by Flycheck, as you suggest.

>  it makes sense to allow individual users to define filters, rather than leaving that responsibility solely with the author of the checker.

Agreed.  Although, if the filtering is somewhat tricky to define, it makes sense to have a default one provided by Flycheck itself.  That way, all users can benefit from this filter, instead of having to carry that in their config.

In particular, how do you intend to extract the relevant part of the error message?  Assuming you have a function `flycheck-filter-haskell-error-message (msg)` which gets the full error message as input and should provide the short error message, how would you write it?

> so a toggle with a keybinding to temporarily disable these filters would also be desirable.

Agreed.  And that means we do not want to to alter the message at the source, but rather give a function that just modifies how the message is displayed.

> It may be possible to implement this with existing Flycheck functionality, without modifying Flycheck itself. 

Well, the error display function is customizable (`flycheck-display-errors-function`), and that is how extensions like `flycheck-inline` change the way they are displayed.  So you could write your own function that filters the error message before displaying it.  Expanding the bodies of the default `flycheck-display-errors-messages` and `flycheck-error-format-message-and-id`, we only have one line to change:

```emacs-lisp
(setq flycheck-display-errors-function
      (lambda (errors)
        (when (and errors (flycheck-may-use-echo-area-p))
          (let ((messages
                 (seq-map
                  (lambda (err)
                    (let ((id (flycheck-error-id err))
                          (filename (flycheck-error-filename err)))
                      (concat (when (and filename (not (equal filename (buffer-file-name))))
                                (format "In \"%s\":\n"
                                        (file-relative-name filename default-directory)))
                              (filter-message ; <-- This is where the filtering happens
                               (flycheck-error-message err))
                              (when id
                                (format " [%s]" id))))))
                 errors)))
          (display-message-or-buffer (string-join messages "\n\n")
                                     flycheck-error-message-buffer
                                     'not-this-window))))
```
But that is not a good solution, since this only changes the behavior of the default display function which prints in the echo area.  To change `flycheck-inline`, you would need to do the same: take the source, expand, add the function call at the right place.

I thought you could use an advice around `flycheck-error-message`, but apparently that doesn't work since this is a macro that gets inlined by the byte-compiler.

So it appears there is no easy way to make that change outside Flycheck.  The change should be simple on our end though: just pass the error message to a customizable filter function.  The downside is that we need to update all extensions that customize the error display to use the filter as well (through a new accessor function to get the potentially-filtered error message, no doubt).

WDYT @cpitclaudel ?

**Comment by cpitclaudel** [comment]
I think this is exactly what error filters are for:

```
(defconst old-filter 
  (flycheck-checker-get 'checker 'error-filter))

(setf (flycheck-checker-get 'old-filter 'error-filter)
      (lambda (errors)
        (setq errors (funcall old-filter errors))
        (seq-do #'some-custom-filter errors)
        errors))
```

Given this, I don't think it's worth adding an extra filter function (sorry for letting this linger so long)
